[nubnubyas] iP

nubnubyas opened the pull request:
# **Monday**

> "Technology is at its best when it brings simplicity and efficiency to our daily lives."

Monday frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- FAST SUPER FAST to use

All you need to do is,

- download the zip from [here](https://github.com/nubnubyas/ip/releases/tag/A-Jar).
- double-click it.
- add your tasks.
- let it manage your tasks for you 😉

And it is **FREE**!

Features:
- [x]  Managing tasks
- [x]  Managing deadlines (coming soon)
- [ ]  Reminders (coming soon)

#
If you Java programmer, you can use it to practice Java too. Here's the main method:
```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

nubnubyas committed:
Completed Level-0

nubnubyas committed:
Completed Level-0

nubnubyas committed:
Completed Level-1 Echo

nubnubyas committed:
Completed Level-2 Add, List

nubnubyas committed:
Completed Level-3 Mark as done

nubnubyas committed:
Completed Level-4 ToDos, Events, Deadlines

nubnubyas committed:
Completed A-TextUiTesting

nubnubyas committed:
Completed Level 5 Handle Errors

nubnubyas committed:
Completed Level 5 Handle Errors

nubnubyas committed:
Completed Level 6 Delete

nubnubyas committed:
Completed A-Enums

nubnubyas committed:
Added Documentations

nubnubyas committed:
Added TestCases

nubnubyas committed:
Add Level-7 Save

nubnubyas committed:
Merge pull request #1 from nubnubyas/branch-Level-7

Merge branch-Level-7

nubnubyas committed:
Add Level-7 Save

nubnubyas committed:
Merge pull request #2 from nubnubyas/branch-Level-7

Merge Level-7 Save

nubnubyas committed:
Add Level-8 Dates and Times

nubnubyas committed:
Merge pull request #3 from nubnubyas/branch-Level-8

Merge Level-8 Dates and Times

nubnubyas committed:
Complete A-MoreOOP

nubnubyas committed:
Merge pull request #4 from nubnubyas/branch-A-MoreOOP

Merge A-MoreOOP

nubnubyas committed:
Complete A-Packages

nubnubyas committed:
Merge pull request #5 from nubnubyas/branch-A-Packages

Merge A-Packages

nubnubyas committed:
Merge pull request #6 from nubnubyas/add-gradle-support

Add gradle support

nubnubyas committed:
Use Gradle to run Duke

nubnubyas committed:
Merge pull request #7 from nubnubyas/branch-A-Gradle

Merge A gradle

nubnubyas committed:
Use Gradle to run Duke

nubnubyas committed:
Merge pull request #8 from nubnubyas/branch-A-Gradle

Use Gradle to run Duke

nubnubyas committed:
Add JUnit Test for ToDo and DateFormatter

nubnubyas committed:
Merge branch 'branch-A-JUnit'

nubnubyas committed:
Merge branch 'master' of https://github.com/nubnubyas/ip

nubnubyas committed:
Add JavaDoc

nubnubyas committed:
Add JavaDoc

nubnubyas committed:
Modify code to follow Java Coding Standard

nubnubyas committed:
Modify code to follow Java Coding Standard

nubnubyas committed:
Add find task function

nubnubyas committed:
Add JavaDoc for functions and classes

nubnubyas committed:
Add find task function

nubnubyas committed:
Modify storage

nubnubyas committed:
Merge branch 'branch-A-JavaDoc'

nubnubyas committed:
Merge branch 'branch-A-CodingStandard'

nubnubyas committed:
Merge branch 'branch-Level-9'

nubnubyas committed:
Add find function to find according to keyword

nubnubyas committed:
Add find function to find according to keyword

nubnubyas committed:
Merge branch 'branch-Level-9'

nubnubyas committed:
Add checkstyle to gradle

HugeNoob commented on src/main/java/monday/monday/exception/MondayExceptions.java:
```diff
@@ -0,0 +1,27 @@
+ * MondayExceptions is a custom exception class for handling
+ * exceptions related to the Monday application.
+ */
+public class MondayExceptions extends Exception {
```
Since the class refers to a single instance of an exception, perhaps it should be named `MondayException` (singular) instead?

HugeNoob commented on src/main/java/monday/Monday.java:
```diff
@@ -0,0 +1,56 @@
+            Ui.printSeparator();
+            try {
+                isRunning = Parser.mondayParser(userInput, taskList);
+            } catch (MondayExceptions e) {
+                System.out.println(e);
+            } catch (NumberFormatException e) {
+                System.out.println("Mark/UnMark number error. " + e.getMessage());
+            } catch (IllegalArgumentException e) {
+                System.out.println("Argument Error: " + e.getMessage());
+            } catch (IndexOutOfBoundsException e) {
+                System.out.println("Index out of Bound Error: " + e.getMessage());
```
Perhaps these println statements could be done through the Ui class instead? I noticed this issue in several other places too.

nubnubyas replied:
That's a good suggestion, I will look into it.

HugeNoob commented on src/main/java/monday/task/TaskList.java:
```diff
@@ -0,0 +1,135 @@
+     *
+     * @param task The task to be added.
+     */
+    public void addToTask(Task task) {
```
Perhaps this should be named `addToList` or `addTask` instead?

nubnubyas replied:
Good catch, did not realise the bad naming.

HugeNoob commented on src/main/java/monday/monday/parser/Parser.java:
```diff
@@ -0,0 +1,147 @@
+     * @throws MondayExceptions if there are errors related to the Monday application
+     * @throws IllegalArgumentException if the user input is in the wrong format
+     */
+    public static boolean mondayParser(String userInput, TaskList taskList) throws MondayExceptions {
```
Perhaps you could simply rename this method to `parse` instead. Currently, you would have to call `Parser.mondayParser(...)`, which seems a bit verbose. Also, method names should sound like verbs, but it currently sounds like a noun.

nubnubyas replied:
I agree with you, I'll fix that, thanks for pointing out.

HugeNoob reviewed (commented):
Code quality is great in general. Verified with checkstyle to be aligning with the course's Java standards.

Side note, is there any reason why you chose to bundle `datetime`, `exception`, `parser`, and `ui` in a `monday` package? It seems a little unintuitive to have a nested package with the exact same name as its parent. I noticed that you had to do `import package.monday.monday.etc...` in many places, which seems a little strange. Perhaps you could rename the nested package or remove it and place the folders in the same level as the rest.

nubnubyas committed:
Add GUI for chatbot

H1410101 commented on src/main/java/monday/storage/Storage.java:
```diff
@@ -0,0 +1,74 @@
+/**
+ * Class for handling the storage of tasks in a file.
+ */
+public class Storage {
```
It seems intuitive to me to rename this to something more specific, eg `TaskArrayListStorage`, to reflect its intention to be a lot more specific than a general object storage (which could work, but would be implemented through a generic, perhaps).

nubnubyas replied:
Thanks for the suggestion, couldn't really think of a good name back then, thanks for providing a alternative naming.

H1410101 commented on src/main/java/monday/storage/Storage.java:
```diff
@@ -0,0 +1,74 @@
+     * @throws IOException if an I/O error occurs while saving the tasks
+     */
+    public void save(ArrayList<Task> tasks) throws IOException {
+        try (ObjectOutputStream output = new ObjectOutputStream(new FileOutputStream(FILEPATH))) {
+            output.writeObject(tasks);
+        }
```
```suggestion
        output.writeObject(tasks);
```
This might be a bit redundant
(p.s. trying the suggestion feature; feel free to ignore / close / reject)

nubnubyas replied:
Good catch, did not realise the redundant try block. Thank you!

H1410101 commented on src/main/java/monday/monday/parser/Parser.java:
```diff
@@ -0,0 +1,140 @@
+     * @throws MondayExceptions if there are errors related to the Monday application
+     * @throws IllegalArgumentException if the user input is in the wrong format
+     */
+    public static String mondayParser(String userInput, TaskList taskList) throws MondayExceptions {
```
like the reviewer before me has mentioned, this isn't phrased as a verb :P

I'm a lot less sure of this, but:
do you think it's a good idea to make this method throw `NumberFormattingException` as well? Your code handles it properly, but I thought it might add a bit of clarity to the types of exceptions expected from this method. 
To some extent, this ambiguity is inherent for the Exception design pattern in general, though, so feel free to disagree.

nubnubyas replied:
You have a point there on the throwing of NumberFormattingException, I will look into adding that too. Thank you :>.

H1410101 commented on src/main/java/monday/task/TaskList.java:
```diff
@@ -0,0 +1,143 @@
+     *
+     * @return Message with the list of all tasks.
+     */
+    public String displayList() {
```
Rather than `display`, which I believe contradicts the purpose of this method (which is to convert to a String representation), I think this is a very natural functionality to override `toString` for.

nubnubyas replied:
That's a good suggestion, I will work on it.

H1410101 commented on src/main/java/monday/task/TaskList.java:
```diff
@@ -0,0 +1,143 @@
+        taskToEdit.markAsDone();
+        save();
+        return Ui.markTask(taskToEdit);
+    }
+
+    /**
+     * Marks a task as not done.
+     *
+     * @param index The index of the task to be marked as not done.
+     * @return Message after marking the task as not done.
+     * @throws IndexOutOfBoundsException If the index is out of range.
+     */
+    public String unMark(int index) {
+        if (index < 1 || index > list.size()) {
+            throw new IndexOutOfBoundsException("Task index is out of range. "
+                    + "Check the number of tasks using the 'list' command.");
+        }
+        Task taskToEdit = list.get(index - 1);
+        taskToEdit.unMark();
+        save();
+        return Ui.unMarkTask(taskToEdit);
+    }
+
+    /**
+     * Deletes a task from the task list.
+     *
+     * @param index The index of the task to be deleted.
+     * @return Message after deleting the task.
+     * @throws IndexOutOfBoundsException If the index is out of range.
+     */
+    public String delete(int index) {
+        if (index < 1 || index > list.size()) {
+            throw new IndexOutOfBoundsException("Task index is out of range. "
+                    + "Check the number of tasks using the 'list' command.");
+        }
+        Task taskToEdit = list.get(index - 1);
+        list.remove(index - 1);
+        save();
+        return Ui.deleteTask(taskToEdit, list.size());
+    }
```
I'm personally not a fan of referring to `Ui` from within `TaskList`; it feels like this could be achieved without "interlocking" classes this way, perhaps by shifting this printing functionality to `Parser`.

H1410101 commented on src/main/java/monday/task/Task.java:
```diff
@@ -0,0 +1,57 @@
+    /**
+     * Marks the task as not done by setting its completion status to false.
+     */
+    public void unMark() {
```
I think there's an opportunity here to mirror the phrasings of related methods.
```suggestion
    public void markAsUndone() {
```

nubnubyas replied:
That's a good suggestion, was feeling that the naming was unnatural too. Thank you!

H1410101 reviewed (approved):
LGTM
Minor nits wrt naming (from previous reviewer as well), and opinionated comments/suggestions from me regarding specific code design choices, but code is quite clean. I like it!

nubnubyas committed:
Add assertions to check task list size.

nubnubyas committed:
Fix coding standard violations in classes

The code in classes had several violations of the Java coding standard
as pointed out in the comments on the pull request. This commit
addresses those issues and ensures that the code now adheres to the
prescribed coding standard.

Changes made:
- Moved println statements to Ui class.
- Renamed MondayExceptions to MondayException.
- Add a TaskList variable to Parser.
- Renamed unMarkTask to markAsUndone in Ui class.
- Renamed Storage to TaskArrayListStorage.
- Removed redundant try block in the function save.
- Renamed addToTask to addToList
- Renamed displayList to toString

This commit improves code readability and maintainability by adhering
to the recommended coding conventions.

nubnubyas committed:
Refactor and improve code structure in Parser.java

This commit refactors the Parser class in the following ways:
- Created separate private methods for each command's logic to prevent
long methods.
- Added a DEFAULT command to handle unrecognized commands gracefully.

These changes make the code more organized, easier to understand, and
better aligned with Java coding standards.

nubnubyas committed:
Add assertions to check task list size

nubnubyas committed:
Merge changes from Main

nubnubyas committed:
Refactor and improve code structure in Parser.java

nubnubyas committed:
Add Continuos Integration

nubnubyas committed:
Merge branch 'master' into branch-A-CI

nubnubyas committed:
Merge branch 'master' into branch-A-CI

nubnubyas committed:
Fix indentation for switch case statement

nubnubyas committed:
Fix indentation for switch case statement

nubnubyas committed:
Merge branch 'master' into branch-A-CI

nubnubyas committed:
Add Continuos Integration

nubnubyas committed:
Add more flexibility in searching for items

nubnubyas committed:
Add more flexibility in searching for items

nubnubyas committed:
Add javadoc for most functions

nubnubyas committed:
Merge branch 'master' of https://github.com/nubnubyas/ip

nubnubyas committed:
Add user guide and clean up for release

nubnubyas committed:
Add user guide and clean up for release

nubnubyas committed:
Add user guide

nubnubyas committed:
Update parser to handle NumberFormatException

nubnubyas committed:
Update error handling for release

nubnubyas committed:
Update User Guide

nubnubyas committed:
Update User Guide

nubnubyas committed:
Update User Guide

nubnubyas committed:
Clean up code for release

nubnubyas committed:
Updated Text Ui Testing

nubnubyas committed:
Cleared file for testing
